core/src/test/java/org/elasticsearch/discovery/zen/ZenDiscoveryTests.java

```diff
@@ -215,4 +216,88 @@ public void handleException(TransportException exp) {
         assertThat(reference.get(), notNullValue());
         assertThat(ExceptionsHelper.detailedMessage(reference.get()), containsString("cluster state from a different master then the current one, rejecting "));
     }
+
+    @Test
+    public void testHandleNodeJoin_incompatibleMinVersion() {
+        Settings nodeSettings = Settings.settingsBuilder()
+                .put("discovery.type", "zen") // <-- To override the local setting if set externally
+                .build();
+        String nodeName = internalCluster().startNode(nodeSettings, Version.V_2_0_0);
+        ZenDiscovery zenDiscovery = (ZenDiscovery) internalCluster().getInstance(Discovery.class, nodeName);
+
+        DiscoveryNode node = new DiscoveryNode("_node_id", new LocalTransportAddress("_id"), Version.V_1_6_0);
+        final IllegalStateException[] e = new IllegalStateException[1];
+        zenDiscovery.handleJoinRequest(node, new MembershipAction.JoinCallback() {
+            @Override
+            public void onSuccess() {
+            }
+
+            @Override
+            public void onFailure(Throwable t) {
+                e[0] = (IllegalStateException) t;
+            }
+        });
+
+        assertThat(e[0], notNullValue());
+        assertThat(e[0].getMessage(), equalTo("Can't handle join request from a node with a version [1.6.0] that is lower than the minimum compatible version [2.0.0-SNAPSHOT]"));
+    }
+
+    @Test
+    public void testHandleNodeJoin_incompatibleMaxVersion() {
+        Settings nodeSettings = Settings.settingsBuilder()
+                .put("discovery.type", "zen") // <-- To override the local setting if set externally
+                .build();
+        String nodeName = internalCluster().startNode(nodeSettings, Version.V_1_6_0);
+        ZenDiscovery zenDiscovery = (ZenDiscovery) internalCluster().getInstance(Discovery.class, nodeName);
+
+        DiscoveryNode node = new DiscoveryNode("_node_id", new LocalTransportAddress("_id"), Version.V_2_0_0);
+        final IllegalStateException[] e = new IllegalStateException[1];
```

Comment:
the reason I used an array here, is because setting/getting happens on the same thread. but I can change it to atomic reference.